### PR TITLE
Move allowsDynamicSplitting to Reader, and set it in CompressedSource.

### PR DESCRIPTION
**Pull request by peihe**
Before the change in CompressedSource, testUnsplittable() fails in:
assertNull(reader.splitAtFraction(fractionConsumed));

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java [hunk 1]
```diff
@@ -583,6 +586,38 @@ public void testGzipProgress() throws IOException {
...
+    BoundedReader<Byte> reader = source.createReader(options);
+
+    List<Byte> actual = Lists.newArrayList();
+    for (boolean hasNext = reader.start(); hasNext; hasNext = reader.advance()) {
+      actual.add(reader.getCurrent());
+      // checkpoint every 9 elements
+      if (actual.size() % 9 == 0) {
+        Double fractionConsumed = reader.getFractionConsumed();
+        if (fractionConsumed != null) {
+          assertNull(reader.splitAtFraction(fractionConsumed));
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: when can fractionConsumed be null? This test might fail for a variety of reasons.

Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: or rather, might silently pass. E.g., if fractionConsumed is always null.

Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: Updated

**Comment by dhalperi** [comment]
LGTM, will merge once green.

**Pull request closed without merging**